### PR TITLE
Implement CompilerDesugaringKind enum

### PR DESCRIPTION
**Pull request by huntiep**
This is the first step outlined in #35946. I think that the variants of `CompilerDesugaringKind` should be changed, I didn't know what the official names for `...` and `<-` are.

I'm not to sure how tests for the compiler work, but I would imagine that tests should be added such that
`Symbol::intern(s) == CompilerDesugaringKind::from(s).as_symbol()` for valid `s`.

### CODE REVIEW

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -397,11 +397,12 @@ impl<'a> LoweringContext<'a> {
     }
 
     fn allow_internal_unstable(&self, reason: &'static str, mut span: Span) -> Span {
```
Comment (oli-obk) [inline src/librustc/hir/lowering.rs]: Wouldn't it make sense to directly pass the enum as an argument here?

**Comment by carols10cents** [comment]
Hi! Thanks for the pull request! We'll be checking in periodically to make sure @nikomatsakis or someone else reviews this.

File: src/libsyntax_pos/hygiene.rs [hunk 2]
```diff
@@ -337,7 +337,39 @@ pub enum ExpnFormat {
...
+        let s = match *self {
+            BackArrow => "<-",
+            DotFill => "...",
+            QuestionMark => "?",
+        };
+        Symbol::intern(s)
+    }
+}
+
+impl<'a> From<&'a str> for CompilerDesugaringKind {
```
Comment (nikomatsakis) [inline src/libsyntax_pos/hygiene.rs]: Somewhat along the lines of what @oli-obk was suggesting, I didn't expect to have `CompiledDesugaringKind::from`, but rather replace all places in the compiler where we use constant strings with direct references to the enum variants, and instead invoke `as_symbol()` anywhere that we need to print things out.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This is in the right direction, but not quite what I had in mind. Please do let me know if my suggestion doesn't make sense or needs clarification!

**Comment by nikomatsakis** [comment]
Also, I apologize for delay, I was on vacation!

**Comment by huntiep** [comment]
It's not a problem, I was also on vacation until recently.

**Review by nikomatsakis — APPROVED** [review]

**Comment by kennytm** [comment]
@bors retry #43283 (`arm-android`)

This seems to become more frequent recently 🤔

**Pull request merged**